# Rice-root-response-atlas

Code and supplementary files used in analysis for "Gene regulatory networks shape developmental plasticity of root cell types under water extremes" (2022)

**Title:** Gene regulatory networks shape developmental plasticity of root cell types under water extremes

**Authors:**  [Mauricio A. Reynoso](https://github.com/reynosoma),* [Alex Borowsky](https://github.com/alexborowsky),* Germain Pauluzzi*, Elaine Yeung, Jianhai Zhang, Elide Formentin, Joel Velasco, Sean Cabanlit, Christine Duvenjian, Matthew J. Prior, Garo Z. Akmakjian, [Roger B. Deal](https://github.com/DealLab), Neelima Sinha, Siobhan M. Brady, [Thomas Girke](https://github.com/tgirke), [Julia Bailey-Serres](https://github.com/jbserres)
_*These authors contributed equally_

**Journal:** 

**Summary:** 

> Understanding how roots modulate development under varied irrigation or rainfall is crucial for development of climate resilient crops. We established a toolbox of tagged rice lines to profile translating mRNAs and chromatin accessibility within specific cell populations. We used these to study roots in a range of environments: plates in the lab, controlled greenhouse stress and recovery conditions, and outdoors in a paddy. Integration of chromatin and mRNA data resolves diverse regulatory networks: cycle genes in proliferating cells that attenuate DNA synthesis under submergence; genes involved in auxin signaling, the circadian clock, and small RNA regulation in ground tissue; and suberin biosynthesis, iron transporters, and nitrogen assimilation in endodermal/exodermal cells modulated with water availability. By applying a systems approach we identify known and candidate driver transcription factors of water deficit responses and xylem development plasticity. Collectively, this resource will facilitate genetic improvements in root systems for optimal climate resilience.



## Code used by topic

- [ATAC-seq mapping THSs definition](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/Mapping_THS_definition_counts_ATAC.R)
- [ATAC-seq THS differential limma-voom & MDS](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/THS_diff_volcano_MDS.R)
- [RNA mapping counting](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/polyA-TRAP-mapping-counting.R)
- [Differential Gene Expression by limma-voom](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/DEG_analysis_limma_polyA_TRAP.R)
- [Clustering](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/blob/master/DEG-analysis-limma-voom/Scripts/Clustering_Heatmap.R)
- [TF enrichment analysis on THS & heatmap](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/THS_enrichment_analysis.R)
- [FIMO motif](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/FIMO_annotated.R)
- [Networks TF to gene target](https://github.com/plant-plasticity/rice-root-response-atlas/blob/main/network_data_frame_production.R)

## Other resources by topic
- [Jbrowser - genome viewer of ATAC-seq, total RNA, TRAP RNA]()
- [Supplemental material](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Cytoscape)

